### PR TITLE
Reader: Filter main feed based on sidebar selection

### PR DESCRIPTION
**Pull request by mehmoodak**
Related to https://github.com/Automattic/loop/issues/236

## Proposed Changes

* Extends the `Recent` menu item on sidebar with the list of subscribed posts which have also made it consistent with the new designs.
* Update the `/following` API calls to include the selected `feed_id`.

## Why are these changes being made?

To enhance the filtering capabilities for feeds within the Reader's "stream" view.

## Testing Instructions

* Navigate to `/read`
* Verify that filtering is working as expected
* Navigate to `/read?flags=reader%2Frecent-feed-overhaul` and select the `stream` view
* Verify that filtering is working as expected

## Pre-merge Checklist

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

File: client/reader/sidebar/reader-sidebar-recent/index.tsx [hunk 1]
```diff
@@ -75,6 +77,17 @@ const ReaderSidebarRecent = ( {
 		if ( ! RECENT_PATH_REGEX.test( path ) ) {
 			page( '/read' );
 		}
+
+		// Analytics.
+		if ( feedId ) {
+			recordAction( 'clicked_reader_sidebar_followed_single_site' );
+			recordGaEvent( 'Clicked Reader Sidebar Followed Single Site' );
+			dispatch( recordReaderTracksEvent( 'calypso_reader_sidebar_followed_single_site_clicked' ) );
```
Comment (mehmoodak) [inline client/reader/sidebar/reader-sidebar-recent/index.tsx]: Added new events for selection on a particular feed. Let me know if there is any feedback on it.

Will open the PR for `calypso_reader_sidebar_followed_single_site_clicked` event registration.

File: client/reader/stream/index.jsx [hunk 2]
```diff
@@ -613,6 +636,18 @@ class ReaderStream extends Component {
 			isRequesting = true;
 		}
 
+		// Due to infinite scrolling or fast selection, API call responses may include items from previously selected feeds.
+		// To temporarily address this issue, filtering out items that don't belong to the currently selected feed.
+		//
+		// The proper fix here is to cancel the pending API calls after the feed selection is changed.
```
Comment (mehmoodak) [inline client/reader/stream/index.jsx]: The current workaround successfully filters content when a specific feed is chosen, but the problem remains when rapidly switching between feeds and ultimately selecting the "All" option.

Are there any ideas on how to properly resolve this issue? I've tried using `AbortController`, but the implementation was taking time. For now I've put that approach on hold and would appreciate feedback if any.

File: client/state/reader-ui/sidebar/selectors.js
Comment (mehmoodak) [inline client/state/reader-ui/sidebar/selectors.js:1]: Migrated to TypeScript.

File: client/state/reader-ui/sidebar/selectors.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+export function isTagsOpen( state: AppState ): boolean {
+	return state.readerUi.sidebar.isTagsOpen;
+}
+
+/**
+ * Get the selected feed ID from the reader sidebar.
+ */
+export function getSelectedFeedId( state: AppState ): number | null {
+	return state.readerUi.sidebar.selectedRecentSite;
+}
```
Comment (mehmoodak) [inline client/state/reader-ui/sidebar/selectors.ts:41]: Just added this function.

File: client/state/reader/analytics/actions.js [hunk 4]
```diff
@@ -2,6 +2,10 @@ import { buildReaderTracksEventProps } from 'calypso/reader/stats';
 import { recordTracksEvent } from 'calypso/state/analytics/actions';
 import { getReaderFollowsCount } from 'calypso/state/reader/follows/selectors';
 
+/**
+ * Record a tracks event with additional reader-specific properties.
+ * @returns The action object to dispatch.
```
Comment (mehmoodak) [inline client/state/reader/analytics/actions.js:6]: To make TS happy.

File: .vscode/launch.json [hunk 5]
```diff
@@ -0,0 +1,40 @@
+{
```
Comment (holdercp) [inline .vscode/launch.json]: This looks like VSCode debug configs. To keep this PR clean, I'd suggest we move this to a separate PR, which I think you have one open already.

File: client/reader/sidebar/reader-sidebar-recent/index.tsx (hunk 1 above)
Comment (holdercp) [reply client/reader/sidebar/reader-sidebar-recent/index.tsx]: I just want to double-check that this PR can be merged before those events are set up, correct?

File: client/reader/sidebar/index.jsx [hunk 6]
```diff
@@ -194,25 +185,13 @@ export class ReaderSidebar extends Component {
...
-				) : (
-					<SidebarItem
-						className={ ReaderSidebarHelper.itemLinkClass( '/read', path, {
-							'sidebar-streams__following': true,
-						} ) }
-						label={ recentLabelTranslationReady ? translate( 'Recent' ) : translate( 'Following' ) }
-						onNavigate={ this.handleReaderSidebarFollowedSitesClicked }
-						customIcon={ <ReaderFollowingIcon viewBox="-3 0 24 24" /> }
-						link="/read"
+				<li className="sidebar-streams__following">
```
Comment (holdercp) [inline client/reader/sidebar/index.jsx:188]: This change will "launch" this feature. We want to keep this flag in place until we're ready to release.

File: client/reader/sidebar/reader-sidebar-recent/index.tsx [hunk 7]
```diff
@@ -75,6 +78,17 @@ const ReaderSidebarRecent = ( {
 		if ( ! RECENT_PATH_REGEX.test( path ) ) {
 			page( '/read' );
 		}
+
+		// Analytics.
+		if ( feedId ) {
+			recordAction( 'clicked_reader_sidebar_followed_single_site' );
+			recordGaEvent( 'Clicked Reader Sidebar Followed Single Site' );
+			recordReaderTracksEvent( 'calypso_reader_sidebar_followed_single_site_clicked' );
```
Comment (holdercp) [inline client/reader/sidebar/reader-sidebar-recent/index.tsx:86]: I fixed the TS errors by converting the recordReaderTracksEvent into a hook so we can use it in this functional component. The original version was meant to be used in redux's `connect()`.

File: client/reader/stream/index.jsx (hunk 2 above)
Comment (holdercp) [reply client/reader/stream/index.jsx]: Ah, yeah that is going to happen since we're sharing the same `following` stream state in Redux. You're right in that I think the right fix is to cancel pending requests each time a new feed is selected in the sidebar. This isn't ideal but I don't really know of another fix with this approach.

@allilevine @DustyReagan  any thoughts here?

File: client/state/reader/analytics/useRecordReaderTracksEvent.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
+import { useDispatch, useSelector } from 'react-redux';
```
Comment (holdercp) [inline client/state/reader/analytics/useRecordReaderTracksEvent.ts:1]: This is just a "hookified" version of the action creator of the same name.

**Review by holdercp — COMMENTED** [review]
Overall I think this looks good and we can roll forward with this.

I did stub out another approach that addresses some of the issues like the feeds overwriting each other if selections are made in quick succession. Here's the [branch](https://github.com/Automattic/wp-calypso/tree/protoype/test-using-feed-endpoint) if you want to take a look. It's just prototype code but might be a direction we could take.

I fixed some TS errors that were related to the analytics calls, and I also suggest we remove the VSCode debug configs from this PR to keep this one tidy.

File: client/reader/stream/index.jsx (hunk 2 above)
Comment (allilevine) [reply client/reader/stream/index.jsx]: I found two alternatives:

1. Assign an id (e.g. a timestamp) to each request in `fetchNextPage`, and then check if we're on the current request when receiving the data. It doesn't cancel the pending request, but it doesn't return the data from the old request.
2. Do something similar with a Redux thunk that "cancels" the older request.

Neither completely cancels the network request like `AbortController` though.

File: .vscode/launch.json (hunk 5 above)
Comment (mehmoodak) [reply .vscode/launch.json]: I've mistakenly commited it. It should be removed now https://github.com/Automattic/wp-calypso/pull/97123/commits/214e072dd9ca2cb573f45c04519195672e7755c6

File: client/reader/sidebar/reader-sidebar-recent/index.tsx (hunk 1 above)
Comment (mehmoodak) [reply client/reader/sidebar/reader-sidebar-recent/index.tsx]: Yes. They will be appear in `rejected` status until we register the event.

File: client/reader/stream/index.jsx (hunk 2 above)
Comment (mehmoodak) [reply client/reader/stream/index.jsx]: ~I will look into this in a separate PR. Let me know if there are any concerns.~

File: client/reader/sidebar/index.jsx (hunk 6 above)
Comment (mehmoodak) [reply client/reader/sidebar/index.jsx:188]: This change is just making the sidebar consistent in both old and new views, it shouldn't have any effect in launching the feature.

We are using `reader/recent-feed-overhaul` flag [here](https://github.com/Automattic/wp-calypso/blob/214e072dd9ca2cb573f45c04519195672e7755c6/client/reader/following/main.tsx#L67) as well which should prevent launching this feature to all customers.

Let me know if I missed something 🙂

File: client/reader/sidebar/reader-sidebar-recent/index.tsx (hunk 7 above)
Comment (mehmoodak) [reply client/reader/sidebar/reader-sidebar-recent/index.tsx:86]: Thanks for creating the hook.

I have improved it further by adding a `dispatchReaderTracksEvent` utility function that ensures consistency between Redux action and React hook over time.

Commit: https://github.com/Automattic/wp-calypso/pull/97123/commits/a372393ad4d7d57ffd21b50cf7f24e2bde86733c

File: client/state/reader/analytics/analytics.utils.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+	pathnameOverride?: string;
+	post: object | null;
+}
+
+/**
+ * A utility function to dispatch a reader specific Tracks event.
+ *
+ * Note: This function shouldn't be used directly, either use `useRecordReaderTracksEvent` hook or `recordReaderTracksEvent` Redux action.
+ */
+export function dispatchReaderTracksEvent(
```
Comment (mehmoodak) [inline client/state/reader/analytics/analytics.utils.ts:19]: Added to keep the implementation consistent between React hook and Redux action.

File: client/reader/stream/index.jsx (hunk 2 above)
Comment (mehmoodak) [reply client/reader/stream/index.jsx]: > Here's the [branch](https://github.com/Automattic/wp-calypso/tree/protoype/test-using-feed-endpoint) if you want to take a look. It's just prototype code but might be a direction we could take.

@holdercp, thank you for investigating this issue. I've found that we are using a same pattern in other parts of the  codebase

Addressed the feedback in 06176c57f08039d99e827af00264b10b076d16cc and removed my temporary fix.

File: client/reader/sidebar/index.jsx (hunk 6 above)
Comment (holdercp) [reply client/reader/sidebar/index.jsx:188]: Ah, I see now! My apologies. Thanks for pointing me to the correct flag conditional.

**Review by holdercp — APPROVED** [review]
This is working well!

- Feed is filtering by selection
- I can't trigger the "overriding feeds" issue 👍
